core/src/main/java/org/elasticsearch/env/NodeEnvironment.java

```diff
@@ -845,6 +846,35 @@ public int getNodeLockId() {
         return shardIds;
     }
 
+    /**
+     * Find all the shards for this index, returning a map of the {@code NodePath} to the number of shards on that path
+     * @param index the index by which to filter shards
+     * @return a map of NodePath to count of the shards for the index on that path
+     * @throws IOException if an IOException occurs
+     */
+    public ObjectLongHashMap<NodePath> shardCountPerPath(final Index index) throws IOException {
+        assert index != null;
+        if (nodePaths == null || locks == null) {
+            throw new IllegalStateException("node is not configured to store local location");
+        }
+        assertEnvIsLocked();
+        final ObjectLongHashMap<NodePath> shardCountPerPath = new ObjectLongHashMap<>();
+        final String indexUniquePathId = index.getUUID();
+        for (final NodePath nodePath : nodePaths) {
+            Path location = nodePath.indicesPath;
+            if (Files.isDirectory(location)) {
+                try (DirectoryStream<Path> indexStream = Files.newDirectoryStream(location)) {
```

Comment:
I wonder if it would be easier to do `if (Files.exists(location.resolve(indexUniquePathId))` then you don't need to iterate?